### DISCUSSION TITLE
How to disable CSS Modules?

### DISCUSSION THREAD

**Discussion by kiril-daskalov** [vercel/next.js#24013] [Help]
Hello!

Please tell me how to disable CSS Modules so I can import CSS/LESS/SASS as good old way like regular React project please!

**Comment by iagobruno** [comment]
I want to know too.

**Comment by pooyarm** [comment]
I also need help here!

Our product is using less for a long time using custom next.config file which is using `@zeit/next-less` and `@zeit/next-css`.
Each react component includes the corresponding less file. Recently and since `@zeit/next-less` and `@zeit/next-css` are deprecated we planned to switch to scss but unfortunately nextjs is forcing me to use css modules and doesn't allow to import scss files except from `_app.jsx`! Refactoring all components at once needs much effort and will lead to UI bugs. I think using scss modules should be optional (of course can be enabled by default).

**Comment by Pulko** [comment]
Having a way to optionally use css modules would be a good feature!

Is there any way to not switch to css modules?

**Comment by zach2825** [comment]
I'd like to chime in and say I HATE the CSS module implementation and because of this GARBAGE, it's making me want to ditch NextJs for all my projects and never recommend this crap. A framework is opinionated that's fine. But, it can't stand in the way of development. If CSS modules are the way to go for this framework then I for one cannot continue to use this once amazing framework. 

I'm all up for moving forward and to the next big thing. But, this is crap. I can't express how much I hate this

**Reply by GabenGar** [reply]
If you hate CSS modules that much, what is the alternative you've expected?

**Reply by jremen** [reply]
A simple import of CSS files anywhere, as is possible with other frameworks or directly in React via regular webpack config.
Not just import into main app.js but in any JS/JSX file.

I think Next is intended for headless web servers. Given the nature of many CMSs, using CSS modules is absolutely out. People who are using CMS with more advanced editors etc. have to work with a regular HTML content coming in, especially if we are talking about articles, posts etc – there is no way to rewrite every new post content as component.

What we are asking for is just a simple option as standard webpack css-loader has: cssModules: true/false/auto

Given the fact on how current webpack config in Next looks like (it's a huge mess) I'm losing my hope.

**Reply by tiborgg** [reply]
@GabenGar believe it or not, but some people and even some teams know how to use CSS properly, as it was intended to be used.

**Reply by cdev-enter** [reply]
An option in the next.config.js to set the default mode for the css loader to global would just do the trick I think

**Reply by zach2825** [reply]
I'm sorry I missed all the replies and completely forgot about this. I have projects where the owner paid for themes, so instead of rewriting all the CSS for those projects, I want to include CSS, and the CSS modules break that. I'd rather be able to use CSS and scss the way I choose.

**Reply by silasbispo** [reply]
CSS Modules é uma desgraça!!!!

**Comment by jremen** [comment]
How can I remove CSS modules support FOREVER? In Next 12.2 webpack config has changed (again) and old trick to remove it via code change does not longer work:
```
config.module.rules[3].oneOf.forEach((one) => {
  if (!`${one.issuer?.and}`.includes('_app')) return;
  one.issuer.and = [path.resolve(__dirname)];
 });
```

I HATE CSS Modules and I CAN'T use it with headless WP as we have a lot of content coming from Elementor etc. in plain HTML so changing CSS classes with module prefix is not possible (and I wouldn't do it anyway).

Please, for christ sake, ALLOW US to turn this garbage off.

**Reply by InNickF** [reply]
I'm not sure what is exactly happening  here  because config as type "any" but this works for me as work around:
module.exports = {
  reactStrictMode: true,
  webpack(config) {
    config.module.rules[2]?.oneOf?.forEach((one) => {
      if (!\${one.issuer?.and}`.includes('_app')) return
      one.issuer.and = [path.resolve(__dirname)]
    })
    return config
  }
}

the exacly change was here:  config.module.rules[ 2 ] ? .oneOf ? .forEach((one) => {

Note: Next.js v12.5.5

**Comment by cwtuan** [comment]
Disable css-module in next.js! 
It is verified in next.js version: v12.3.1 (Sept, 2022)
Updated: As @jremen mentioned, this solution should also works in next.js@13

In the next.config.js, add the following config:
```js
webpack(config) {
    //  Source: https://github.com/vercel/next.js/discussions/24013#discussioncomment-3787335
    config.module.rules.forEach((rule) => {
      const { oneOf } = rule;
      if (oneOf) {
        oneOf.forEach((one) => {
          if (!`${one.issuer?.and}`.includes('_app')) return;
          one.issuer.and = [path.resolve(__dirname)];
        });
      }
    })
    return config;
  },
```

**Reply by zach2825** [reply]
Unfortunately this issue caused a deal breaker for NextJS for my project :(

**Reply by MorganeLeCaignec** [reply]
Great, thank you!
I'm not too familiar with webpack config, does anyone know if this could have unwanted consequences somewhere else or does it only remove the importing of css/scss outside of _app?

**Reply by cwtuan** [reply]
Here is a [json to typescript tool](https://json-5.com/jsontots) built by next.js without css-modules. No unwanted side effect so far.

See the result html:
![image](https://user-images.githubusercontent.com/3455798/196075726-a8572662-4ec4-4e25-a232-0ea0701ad1ae.png)

**Reply by jremen** [reply]
To me it doesn't have any consequences. Problem is, it changes with Webpack config changes in newer Next.js versions and in newer ones it doesn't work anymore.

There should be a real systematic approach build into Next.js itself as these are just mere hacks to webpack config

**Comment by jremen** [comment]
Any news about this regarding just released Next.js 13.0?

**Reply by jremen** [reply]
Just upgraded project to Next 13, apart from the fact that I had to remove all ` <a>` elements from `<Link>` component (and move the props to the `<Link>`) it seems that solution from https://github.com/vercel/next.js/discussions/24013#discussioncomment-3787335 works as it used to for Next 12.3.1

**Comment by teodragovic** [comment]
Hello. I'm in the middle of migrating from `preact-cli` to Next v12 and I had this same issue. This is custom webpack config that worked for me:

```js
    webpack: (config, { dev, isServer }) => {
        config.plugins.push(
            new MiniCssExtractPlugin({
                filename: !dev ? 'static/css/[name].[contenthash:5].css' : 'static/css/[name].css',
                chunkFilename: !dev
                    ? 'static/css/[name].chunk.[contenthash:5].css'
                    : 'static/css/[name].chunk.css'
            })
        );
        config.module.rules.push({
            test: /\.scss$/,
            use: [
                (isServer || !dev) ? MiniCssExtractPlugin.loader : require.resolve('style-loader'),
                {
                    loader: require.resolve('css-loader'),
                    options: {
                        modules: false, // <--- disable CSS modules
                        importLoaders: 2,
                        sourceMap: true
                    }
                },
                {
                    loader: require.resolve('postcss-loader'),
                    options: {
                        sourceMap: true,
                        postcssOptions: {
                            plugins: [autoprefixer()]
                        }
                    }
                },
                {
                    loader: require.resolve('sass-loader'),
                    options: {
                        sourceMap: true
                    }
                }
            ]
        });
        return config;
    },
```

The config is pretty much copied from `preact-cli`. We use Sass and import global stylesheet. I wanted to preserve that setup but be able to add it on page level (to separate user app from marketing pages etc.). Next docs say any custom CSS config disables their built-in CSS support so no need to pluck out existing rules. Once I explicitly disabled modules in `css-loaders` options, everything started to work as expected.

**Comment by mmvsk** [comment]
Here is a workaround that works for me.

My structure: one React component per file (e.g. `button.tsx`), and one stylesheet per component (e.g. `styles/button.scss`). Each component have an unique class (e.g. `.Button`). Children elements use locally scoped classes using direct selectors (e.g. `.Button > .icon`, `.MainLayout > .header > .nav`, ...) and modifiers (`.Button.--primary`, `.List > ... > .row.--selected`).

It works very well, and CSS modules basically solve a problem I don't have, while getting in my way.

The workaround:

1. Define the stylesheet in a (S)CSS module, with a `:global` scope for the children (that prevents contents from being prefixed):

```scss
// styles/button.module.scss

.Button {
    &:global {
        > .label { color: red }
    }
}
```

2. Apply the CSS module class name to the component's root element:

```tsx
// button.tsx

import Styles from "./styles/button.module.scss";

export function Button(props: Props) {
    return (
        <button className={Styles.Button}>
            <span className="label">Button label</span>
        </button>
    );
}
```

Modifiers can be joined directly to the CSS-module-prefixed class name, as long as they are under `:global`:

```ts
[Styles.ComponentName, ...modifiers].join(" ")
```

In the resulting CSS, the component root class will be prefixed, but not its children.

`.button_Button__xyzAB > .label`, where the

**Comment by ricmello** [comment]
I customized next's getLocalIdent function to remove the hash part and reassigned it. it's working so far:

```js
webpack: (config) => {
    const regexEqual = (x, y) =>
      x instanceof RegExp &&
      y instanceof RegExp &&
      x.source === y.source &&
      x.global === y.global &&
      x.ignoreCase === y.ignoreCase &&
      x.multiline === y.multiline

    const cssLoader = config.module.rules
      .find((rule) => typeof rule.oneOf === "object")
      .oneOf.find((rule) => regexEqual(rule.test, /\.module\.(scss|sass)$/))
      .use.find((rule) => rule.loader.includes("css-loader"))

    cssLoader.options = {
      ...cssLoader.options,
      modules: {
        ...cssLoader.options.modules,
        getLocalIdent: (context, _, exportName, options) => {
          const relativePath = path
            .relative(context.rootContext, context.resourcePath)
            .replace(/\\+/g, "/")

          const fileNameOrFolder =
            /(?<!pages[\\/])index\.module\.(scss|sass|css)$/.test(
              relativePath
            )
              ? "[folder]"
              : "[name]"

          return loaderUtils
            .interpolateName(
              context,
              `${fileNameOrFolder}_${exportName}`,
              options
            )
            .replace(/\.module_/, "_")
            .replace(/[^a-zA-Z0-9-_]/g, "_")
            .replace(/^(\d|--|-\d)/, "__$1")
        }
      }
    }

    return config
  }
```

**Comment by NICK-XoX** [comment]
I found ```next@13.0.6``` works as expected for dev mode.
Then after dev, use the lastest version to build and it works as expected.
Let me know if anyone finds a better fix to use the lastest versions on dev.

**Comment by SouthLink** [comment]
I created a plugin to solve this problem , [next-disable-global-css](https://www.npmjs.com/package/next-disable-global-css), support disabling the node_module global css introduction for the app directory

**Reply by Staslios** [reply]
This plugin forces me to create a custom configuration from scatch for my styles, am I wrong ?

**Reply by SouthLink** [reply]
> This plugin forces me to create a custom configuration from scatch for my styles, am I wrong ?

The plugin only disables webpack's global css of next.js and modifies the CSS-loader configuration to achieve its function, but does not add custom configuration. Is there any specific problem description.

**Comment by tomcatmwi** [comment]
Just chiming in to add whoever came up with SCSS modules, and made it mandatory, not only didn't have his thinking caps on, but should be permanently banned from using computers. It wrecked our entire project, and there's no way to salvage it. We must completely rewrite the entire stylesheet system, and it caused our company a LOT of damage.
